### PR TITLE
Add package `@preact/signals-react-transform`

### PR DESCRIPTION
**Pull request by andrewiggins**
This PR adds a new package `@preact/signals-react-transform` that provides a Babel plugin to transform functions to observer/track what signals are used when the function runs. This plugin is the start to an alternate React integration for signals that does not hook into React internals. Once our approach here is finalized and published, I'll update the `@preact/signals-react` package to reference this and provide various options for using signals in React.

Check [the test file](https://github.com/preactjs/signals/pull/375/files#diff-95ee027f0e90ee403bb02268d8b0de381d341a38fd0632ac1ec3838a43bbf365R66-R77) for examples of what the transformation looks like.

I'm currently exploring two different kinds of transformations in this implementation to see which one works better. One just adds a call to `useSignals()` to the top of the component and relies on the next invocation of `useSignals` or a microtick to stop tracking signals. The other wraps the component body in a `try/finally` to start and stop tracking signals. I'm currently leaning towards making the `try/finally` the default cuz it leaves less room for tricky edge cases that may lead to non-obvious behavior and rerenders.

However, the implementation of `useSignals` is the same either way. We'll always setup the microtick to clean up any outstanding effects to prevent any loose ends regardless of which transformation is used. So users can manually  drop `useSignals` into a component and get basic functionality.

### CODE REVIEW

File: karma.conf.js [hunk 1]
```diff
@@ -283,7 +287,7 @@ module.exports = function (config) {
 			jsx: "preserve",
 
 			// esbuild options
-			target: downlevel ? "es5" : "es2015",
+			target: downlevel ? "es5" : "es2019",
```
Comment (andrewiggins) [inline karma.conf.js:290]: Whoops - will undo this. Did this to not downlevel async/await while debugging tests

File: packages/react-transform/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
+# Signals React Transform
```
Comment (andrewiggins) [inline packages/react-transform/README.md:1]: Unsure what's the right level of detail here. Would love other's thoughts.

File: packages/react-transform/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+	PluginObj,
+	PluginPass,
+	NodePath,
+	template,
+} from "@babel/core";
+import { isModule, addNamed, addNamespace } from "@babel/helper-module-imports";
+
+// TODO:
+// - how to trigger rerenders on attributes change if transform never sees
+//   `.value`?
```
Comment (andrewiggins) [inline packages/react-transform/src/index.ts:13]: I still need to build the signal attribute handling in our jsx handler

File: packages/react-transform/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+					);
+				},
+			},
+
+			ArrowFunctionExpression: {
+				// TODO: Consider alternate implementation, where on enter of a function
+				// expression, we run our own manual scan the AST to determine if the
+				// function uses signals and is a component. This manual scan once upon
+				// seeing a function would probably be faster than running an entire
+				// babel pass with plugins on components twice.
```
Comment (andrewiggins) [inline packages/react-transform/src/index.ts:214]: For anyone with more experience with babel transforms - if I want to know if a function body contains JSX to know whether to perform my transformation, is it better to do a local visitor pass inside this ArrowFunctionExpression handler to observe that (only running my local visitor), and then transform? Or to do the observation in the larger babel visitor pass and then transform after, which triggers and additional visitor pass with all plugins?

Comment (marvinhagemeister) [reply packages/react-transform/src/index.ts:214]: That's a good question. I've never thought about this.

File: packages/react-transform/README.md (hunk 2 above)
Comment (marvinhagemeister) [reply packages/react-transform/README.md:1]: I think it's perfectly fine. We could theoretically shorten the intro paragraph about what signals are by just pointing to the blog post. But then again that I think leaving as is, is fine too.

**Comment by andrewiggins** [comment]
TODO for myself: Add some E2E babel transform tests, i.e. transform code and run it.

**Comment by XantreGodlike** [comment]
What if someone calls `useSignals` twice? Or babel will add additional one?

**Comment by XantreGodlike** [comment]
What about cases when signals used inside hooks that component uses? How to handle this case?

**Comment by andrewiggins** [comment]
> What if someone calls useSignals twice? Or babel will add additional one?

**Comment by andrewiggins** [comment]
> What if someone calls useSignals twice? Or babel will add additional one?
`useSignals` can be called multiple times in a component. It'll recreate the effect and other relevant tracking machinery but that's okay. The previous effect will be cleaned up and any signals used after the additional calls to useSignals will still trigger rerenders properly.

> What about cases when signals used inside hooks that component uses? How to handle this case?
Hooks that use signals are transformed to call `useSignals` at the top of the function. In this situation, we are relying on the fact that calling `useSignals` multiple times is safe to do.

File: packages/react-transform/src/index.ts (hunk 3 above)
Comment (andrewiggins) [reply packages/react-transform/src/index.ts:13]: Will do this in a follow up PR. this one is big enough 😅

File: packages/react-transform/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+function hasLeadingOptInComment(path: NodePath) {
+	return hasLeadingComment(path, optInCommentIdentifier);
+}
+
+function hasLeadingOptOutComment(path: NodePath) {
+	return hasLeadingComment(path, optOutCommentIdentifier);
+}
+
+function isOptedIntoSignalTracking(path: NodePath | null): boolean {
```
Comment (andrewiggins) [inline packages/react-transform/src/index.ts:91]: Here and in `isOptedOutOfSignalTracking`, I'm climbing up the AST to find the most logical place for a comment to live. For example supporting a use case such as:
```js
/** @trackSignals */
export default function useCustomHook() {}
```

**Review by andrewiggins — COMMENTED** [review]
Okay! I believe this is now ready for review :tada:

File: packages/react-transform/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+		default:
+			return false;
+	}
+}
+
+function isComponentFunction(path: NodePath<FunctionLike>): boolean {
+	return (
+		fnNameStartsWithCapital(path) && // Function name indicates it's a component
+		getData(path.scope, containsJSX) === true && // Function contains JSX
+		path.scope.parent === path.scope.getProgramParent() // Function is top-level
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply packages/react-transform/src/index.ts:137]: Yea, it's true. I'm thinking HoC will need to manually opt-in using `/** @trackSignals */` or by calling `useSignals()` directly in their wrapper function.

It is harder to detect HoCs and know where to put the useSignals call so for now I'm explicitly skipping them. I should add a note to the Readme calling this out for people who want to use signals in higher order components.

[the remark this answers is not in the document]
Comment (XantreGodlike) [reply packages/react-transform/src/index.ts:137]: I think it a good idea to mention edge cases in readme

File: packages/react-transform/test/node/index.test.tsx [hunk 7]
```diff
@@ -0,0 +1,835 @@
...
+
+			const expectedOutput = `
+				import { useSignals as _useSignals } from "@preact/signals-react/runtime";
+				const MyComponent = () => {
+					var _effect = _useSignals();
+					try {
+						signal.value;
+						return <div>Hello World</div>;
+					} finally {
+						_effect.f();
```
Comment (andrewiggins) [inline packages/react-transform/test/node/index.test.tsx:75]: Any opinions on the name of this method `f`? It's short for `finishEffect` to stop the current effect.

I've changed the return of `useSignals` from an method to an object that can hang other stuff of like the dispose symbol for using with `using` which this PR also includes support for.

File: packages/react-transform/src/index.ts (hunk 4 above)
Comment (JoviDeCroock) [reply packages/react-transform/src/index.ts:339]: You can use something like [this](https://github.com/preactjs/prefresh/blob/main/packages/babel/src/index.mjs#L168-L183) on a VariableDeclaration to see whether they are used as JSX

**Review by JoviDeCroock — APPROVED** [review]
This looks amazing

File: packages/react-transform/src/index.ts (hunk 4 above)
Comment (andrewiggins) [reply packages/react-transform/src/index.ts:339]: Oh interesting. IIUC, this code looks to see if a particular identifier is used as JSX somewhere in the same file?

Comment (JoviDeCroock) [reply packages/react-transform/src/index.ts:339]: ScopeBindings should work across files afaik

**Review by marvinhagemeister — APPROVED** [review]
